### PR TITLE
r/emr studio - new resource

### LINKED ISSUE #20597: Amazon EMR Studio
### Community Note

* Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* Please do not leave "+1" or other comments that do not add relevant new information or questions, they generate extra noise for issue followers and do not help prioritize the request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment

### Description

Allow terraform to create and manage emr studio resources.

### New or Affected Resource(s)

* `aws_emr_studio`

### Potential Terraform Configuration

```hcl
resource "aws_emr_studio" "example" {
  name = "example-studio"
  auth_mode = "SSO"
  vpc_id = <example-vpc-id>
  subnet_ids = ["<subnet-id-1>", "<subnet-id-2>"]
  service_role = <example-studio-service-role-name>
  user-role = <example-studio-user-role-name>
  workspace-security-group-id = <example-workspace-sg-id>
  engine-security-group-id = <example-engine-sg-id>
  default-s3-location = <example-s3-location>
}
```

### References

* https://docs.aws.amazon.com/emr/latest/ManagementGuide/emr-studio.html

### PR DESCRIPTION
**Pull request by DrFaust92**
### Community Note

* Please vote on this pull request by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original pull request comment to help the community and maintainers prioritize this request
* Please do not leave "+1" or other comments that do not add relevant new information or questions, they generate extra noise for pull request followers and do not help prioritize the request

Closes #20597.

Output from acceptance testing:

```
$ make testacc TESTS=TestAccEMRStudio_ PKG=emr

--- PASS: TestAccEMRStudio_disappears (79.81s)
--- PASS: TestAccEMRStudio_sso (85.61s)
--- PASS: TestAccEMRStudio_tags (201.09s)
--- PASS: TestAccEMRStudio_iam (109.14s)
```

### CODE REVIEW

**Review by ewbankkit — APPROVED** [review]
LGTM :rocket:.

##### AWS Commercial account with SSO enabled

```console
% make testacc TESTS=TestAccEMRStudio_ PKG=emr
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test ./internal/service/emr/... -v -count 1 -parallel 20 -run='TestAccEMRStudio_' -timeout 180m
=== RUN   TestAccEMRStudio_sso
=== PAUSE TestAccEMRStudio_sso
=== RUN   TestAccEMRStudio_iam
=== PAUSE TestAccEMRStudio_iam
=== RUN   TestAccEMRStudio_disappears
=== PAUSE TestAccEMRStudio_disappears
=== RUN   TestAccEMRStudio_tags
=== PAUSE TestAccEMRStudio_tags
=== CONT  TestAccEMRStudio_sso
=== CONT  TestAccEMRStudio_tags
=== CONT  TestAccEMRStudio_disappears
=== CONT  TestAccEMRStudio_iam
--- PASS: TestAccEMRStudio_disappears (41.06s)
--- PASS: TestAccEMRStudio_iam (42.37s)
--- PASS: TestAccEMRStudio_sso (67.39s)
--- PASS: TestAccEMRStudio_tags (85.51s)
PASS
ok  	github.com/hashicorp/terraform-provider-aws/internal/service/emr	89.000s
```

##### AWS Commercial account without SSO enabled

```console
% make testacc TESTS=TestAccEMRStudio_ PKG=emr
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test ./internal/service/emr/... -v -count 1 -parallel 20 -run='TestAccEMRStudio_' -timeout 180m
=== RUN   TestAccEMRStudio_sso
=== PAUSE TestAccEMRStudio_sso
=== RUN   TestAccEMRStudio_iam
=== PAUSE TestAccEMRStudio_iam
=== RUN   TestAccEMRStudio_disappears
=== PAUSE TestAccEMRStudio_disappears
=== RUN   TestAccEMRStudio_tags
=== PAUSE TestAccEMRStudio_tags
=== CONT  TestAccEMRStudio_sso
=== CONT  TestAccEMRStudio_tags
=== CONT  TestAccEMRStudio_disappears
=== CONT  TestAccEMRStudio_iam
=== CONT  TestAccEMRStudio_tags
    acctest.go:960: skipping test for aws/us-west-2: Error running apply: exit status 1
        
        Error: error creating EMR Studio: InvalidRequestException: SSO is not enabled (Service: AWSEditors; Status Code: 400; Error Code: InvalidRequestException; Request ID: 8282038c-c3a4-40da-b092-3ac8114434ae; Proxy: null)
        {
          RespMetadata: {
            StatusCode: 400,
            RequestID: "19c33a06-7189-4c73-a979-ef3a92c0e2fc"
          },
          Message_: "SSO is not enabled (Service: AWSEditors; Status Code: 400; Error Code: InvalidRequestException; Request ID: 8282038c-c3a4-40da-b092-3ac8114434ae; Proxy: null)"
        }
        
          with aws_emr_studio.test,
          on terraform_plugin_test.tf line 74, in resource "aws_emr_studio" "test":
          74: resource "aws_emr_studio" "test" {
        
=== CONT  TestAccEMRStudio_disappears
    acctest.go:960: skipping test for aws/us-west-2: Error running apply: exit status 1
        
        Error: error creating EMR Studio: InvalidRequestException: SSO is not enabled (Service: AWSEditors; Status Code: 400; Error Code: InvalidRequestException; Request ID: 8adde1ba-85dd-4054-8e65-46cbe40ba0c7; Proxy: null)
        {
          RespMetadata: {
            StatusCode: 400,
            RequestID: "5eed9ffd-9953-4954-8445-b932a91e83ef"
          },
          Message_: "SSO is not enabled (Service: AWSEditors; Status Code: 400; Error Code: InvalidRequestException; Request ID: 8adde1ba-85dd-4054-8e65-46cbe40ba0c7; Proxy: null)"
        }
        
          with aws_emr_studio.test,
          on terraform_plugin_test.tf line 74, in resource "aws_emr_studio" "test":
          74: resource "aws_emr_studio" "test" {
        
=== CONT  TestAccEMRStudio_sso
    acctest.go:960: skipping test for aws/us-west-2: Error running apply: exit status 1
        
        Error: error creating EMR Studio: InvalidRequestException: SSO is not enabled (Service: AWSEditors; Status Code: 400; Error Code: InvalidRequestException; Request ID: 1cf177cc-1be3-4817-ab7d-026f4071fc64; Proxy: null)
        {
          RespMetadata: {
            StatusCode: 400,
            RequestID: "963bab2d-ef92-4b89-8792-58a66c51ba2b"
          },
          Message_: "SSO is not enabled (Service: AWSEditors; Status Code: 400; Error Code: InvalidRequestException; Request ID: 1cf177cc-1be3-4817-ab7d-026f4071fc64; Proxy: null)"
        }
        
          with aws_emr_studio.test,
          on terraform_plugin_test.tf line 74, in resource "aws_emr_studio" "test":
          74: resource "aws_emr_studio" "test" {
        
--- SKIP: TestAccEMRStudio_tags (28.37s)
--- SKIP: TestAccEMRStudio_disappears (28.38s)
--- SKIP: TestAccEMRStudio_sso (28.44s)
--- PASS: TestAccEMRStudio_iam (42.65s)
PASS
ok  	github.com/hashicorp/terraform-provider-aws/internal/service/emr	45.931s
```

##### US GovCloud

```console
% make testacc TESTS=TestAccEMRStudio_ PKG=emr                
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test ./internal/service/emr/... -v -count 1 -parallel 20 -run='TestAccEMRStudio_' -timeout 180m
=== RUN   TestAccEMRStudio_sso
=== PAUSE TestAccEMRStudio_sso
=== RUN   TestAccEMRStudio_iam
=== PAUSE TestAccEMRStudio_iam
=== RUN   TestAccEMRStudio_disappears
=== PAUSE TestAccEMRStudio_disappears
=== RUN   TestAccEMRStudio_tags
=== PAUSE TestAccEMRStudio_tags
=== CONT  TestAccEMRStudio_sso
=== CONT  TestAccEMRStudio_tags
=== CONT  TestAccEMRStudio_disappears
=== CONT  TestAccEMRStudio_iam
    acctest.go:960: skipping test for aws-us-gov/us-gov-west-1: Error running apply: exit status 1
        
        Error: error creating EMR Studio: ValidationException: Account is not whitelisted to use this feature.
        	status code: 400, request id: e4f12c9a-4fed-497b-a445-1249f4c0102c
        
          with aws_emr_studio.test,
          on terraform_plugin_test.tf line 74, in resource "aws_emr_studio" "test":
          74: resource "aws_emr_studio" "test" {
        
=== CONT  TestAccEMRStudio_sso
    acctest.go:960: skipping test for aws-us-gov/us-gov-west-1: Error running apply: exit status 1
        
        Error: error creating EMR Studio: ValidationException: Account is not whitelisted to use this feature.
        	status code: 400, request id: 957b6542-2bf6-435f-9c1e-eb3f49bfcb7b
        
          with aws_emr_studio.test,
          on terraform_plugin_test.tf line 74, in resource "aws_emr_studio" "test":
          74: resource "aws_emr_studio" "test" {
        
=== CONT  TestAccEMRStudio_tags
    acctest.go:960: skipping test for aws-us-gov/us-gov-west-1: Error running apply: exit status 1
        
        Error: error creating EMR Studio: ValidationException: Account is not whitelisted to use this feature.
        	status code: 400, request id: 1f6b5210-a2e6-4119-8b9d-b9648e6369ac
        
          with aws_emr_studio.test,
          on terraform_plugin_test.tf line 74, in resource "aws_emr_studio" "test":
          74: resource "aws_emr_studio" "test" {
        
=== CONT  TestAccEMRStudio_disappears
    acctest.go:960: skipping test for aws-us-gov/us-gov-west-1: Error running apply: exit status 1
        
        Error: error creating EMR Studio: ValidationException: Account is not whitelisted to use this feature.
        	status code: 400, request id: 60088379-ab72-4da7-b23c-6d2ef6425160
        
          with aws_emr_studio.test,
          on terraform_plugin_test.tf line 74, in resource "aws_emr_studio" "test":
          74: resource "aws_emr_studio" "test" {
        
--- SKIP: TestAccEMRStudio_sso (19.37s)
--- SKIP: TestAccEMRStudio_iam (19.45s)
--- SKIP: TestAccEMRStudio_tags (19.45s)
--- SKIP: TestAccEMRStudio_disappears (19.48s)
PASS
ok  	github.com/hashicorp/terraform-provider-aws/internal/service/emr	23.010s
```

**Comment by ewbankkit** [comment]
@DrFaust92 Thanks for the contribution :tada: :clap:.
In general everything was great.
I modified some of the acceptance tests configurations and error checking.

File: internal/service/emr/managed_scaling_policy_test.go [hunk 1]
```diff
@@ -15,12 +15,24 @@ import (
...
 
+func init() {
+	acctest.RegisterServiceErrorCheckFunc(emr.EndpointsID, testAccErrorCheckSkipEMR)
+}
+
+func testAccErrorCheckSkipEMR(t *testing.T) resource.ErrorCheckFunc {
+	return acctest.ErrorCheckSkipMessagesContaining(t,
+		"Managed scaling is not available",
+		"SSO is not enabled",
+		"Account is not whitelisted to use this feature",
```
Comment (AdamTylerLynch) [inline internal/service/emr/managed_scaling_policy_test.go:26]: AWS has deprecated the use of the terms "whitelist" and "blacklist" in favor of "allow-list" and "deny-list".

File: internal/service/emr/studio.go [hunk 2]
```diff
@@ -0,0 +1,274 @@
...
+
+func resourceStudioDelete(d *schema.ResourceData, meta interface{}) error {
+	conn := meta.(*conns.AWSClient).EMRConn
+
+	request := &emr.DeleteStudioInput{
+		StudioId: aws.String(d.Id()),
+	}
+
+	log.Printf("[INFO] Deleting EMR Studio: %s", d.Id())
+	_, err := conn.DeleteStudio(request)
```
Comment (AdamTylerLynch) [inline internal/service/emr/studio.go:264]: Should there be a waiter here?

File: internal/service/emr/studio.go [hunk 3]
```diff
@@ -0,0 +1,274 @@
...
+		if tfawserr.ErrMessageContains(err, emr.ErrCodeInvalidRequestException, "Service role does not have permission to access") {
+			return resource.RetryableError(err)
+		}
+		if err != nil {
+			return resource.NonRetryableError(err)
+		}
+		return nil
+	})
+	if tfresource.TimedOut(err) {
+		result, err = conn.CreateStudio(input)
```
Comment (AdamTylerLynch) [inline internal/service/emr/studio.go:164]: Is there a reason you decided to not use CreateStudioWithContext in this new resource? This is just a question.

Comment (DrFaust92) [reply internal/service/emr/studio.go:164]: nope, just inertia from previous PRs. need to start doing that.

File: internal/service/emr/studio.go (hunk 2 above)
Comment (DrFaust92) [reply internal/service/emr/studio.go:264]: doesnt look like studio exposes any status attribute so no, also the post destroy checks pass

**Pull request merged**